packages/flutter/lib/src/rendering/wrap.dart

```diff
@@ -0,0 +1,358 @@
+                 .map/*<double>*/((RenderBox child) => child.getMinIntrinsicHeight(double.INFINITY))
+                 .reduce((double value, double element) => math.max(value, element));
+    }
+    return null;
+  }
+
+  @override
+  double computeMaxIntrinsicHeight(double width) {
+    switch (_direction) {
+      case Axis.horizontal:
+        return _computeIntrinsicHeightForWidth(width);
+      case Axis.vertical:
+        return _childIterator
+                 .map/*<double>*/((RenderBox child) => child.getMaxIntrinsicHeight(double.INFINITY))
+                 .reduce((double value, double element) => value + element);
+    }
+    return null;
+  }
+
+  @override
+  double computeDistanceToActualBaseline(TextBaseline baseline) {
+    return defaultComputeDistanceToHighestActualBaseline(baseline);
+  }
+
+  double _getCrossSize(RenderBox child) {
+    switch (_direction) {
+      case Axis.horizontal:
+        return child.size.height;
+      case Axis.vertical:
+        return child.size.width;
+    }
+    return null;
+  }
+
+  double _getMainSize(RenderBox child) {
+    switch (_direction) {
+      case Axis.horizontal:
+        return child.size.width;
+      case Axis.vertical:
+        return child.size.height;
+    }
+    return null;
+  }
+
+  @override
+  void performLayout() {
+    BoxConstraints childConstraints;
+    double maxExtent;
+    // TODO(ianh): Make the asserts below explain that there's no paint putting a Wrap in an unconstrained main axis (since then it's just a poor man's Flex).
+    switch (_direction) {
+      case Axis.horizontal:
+        assert(constraints.maxWidth.isFinite);
+        childConstraints = new BoxConstraints(maxWidth: constraints.maxWidth);
+        maxExtent = constraints.maxWidth;
+        break;
+      case Axis.vertical:
+        assert(constraints.maxHeight.isFinite);
+        childConstraints = new BoxConstraints(maxHeight: constraints.maxHeight);
+        maxExtent = constraints.maxHeight;
+        break;
+    }
+    assert(childConstraints != null);
+    assert(maxExtent != null);
+    RenderBox child = firstChild;
+    List<RenderBox> row = <RenderBox>[];
+    double extent = 0.0;
+    double maxCrossExtent = 0.0;
+    Size maxSize = Size.zero;
+    while (child != null) {
+      child.layout(childConstraints, parentUsesSize: true);
+      double childExtent = _getMainSize(child);
+      if (extent + childExtent > maxExtent) {
+        assert(row.isNotEmpty);
+        maxSize = _fillRow(row, extent, maxExtent, maxCrossExtent, maxSize);
+        row.clear();
+        extent = childExtent;
+        maxCrossExtent = _getCrossSize(child);
+      } else {
+        extent += childExtent;
+        maxCrossExtent = math.max(maxCrossExtent, _getCrossSize(child));
+      }
+      row.add(child);
+      child = childAfter(child);      
+    }
+    if (row.isNotEmpty)
+      maxSize = _fillRow(row, extent, maxExtent, maxCrossExtent, maxSize);
+    size = constraints.constrain(maxSize);
```

Comment:
When you apply these constraints, you might end up with a cross axis extent of the wrap that's bigger than max size.  You might want a property to describe how that free space is allocated.

Also, do you want to do something if we overflow in the cross axis?  For example, perhaps we should clip the visual overflow.